### PR TITLE
kubeadm: Remove ClusterConfiguration from InitConfiguration in v1beta2

### PR DESCRIPTION
**Pull request by rosti**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

Ever since v1alpha3, `InitConfiguration` is containing `ClusterConfiguration` embedded in it. This was done to mimic the internal `InitConfiguration`, which in turn is used throughout the kubeadm code base as if it is the old `MasterConfiguration` of v1alpha2.

This, however, is confusing to users who vendor in kubeadm as the embedded `ClusterConfiguration` inside `InitConfiguration` is not marshaled to YAML.
For this to happen, special care must be taken for the `ClusterConfiguration` field to marshaled separately.

Thus, to make things smooth for users and to reduce third party exposure to technical debt, this change removes `ClusterConfiguration` embedding from `InitConfiguration`.

Doing this would also mean, that the internal config types refactoring can be now decoupled from the v1beta2 config release schedule.

**Which issue(s) this PR fixes**:

Refs kubernetes/enhancements#970 kubernetes/kubeadm#1439

**Special notes for your reviewer**:

Most of this deals with extracting `ClusterConfiguration` from `InitConfiguration` in kubeadm argument passing and test code.

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/priority important-longterm
/assign @fabriziopandini
/assign @timothysc
/assign @neolit123

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: v1beta2 InitConfiguration no longer embeds ClusterConfiguration it it.
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/config.go [hunk 1]
```diff
@@ -176,17 +176,17 @@ func getSupportedComponentConfigAPIObjects() []string {
...
-		},
-		BootstrapTokens: []kubeadmapiv1beta2.BootstrapToken{placeholderToken},
+		BootstrapTokens:  []kubeadmapiv1beta2.BootstrapToken{placeholderToken},
 		NodeRegistration: kubeadmapiv1beta2.NodeRegistrationOptions{
 			CRISocket: constants.DefaultDockerCRISocket, // avoid CRI detection
 		},
-	})
+	}
+	clusterCfg := &kubeadmapiv1beta2.ClusterConfiguration{
+		KubernetesVersion: fmt.Sprintf("v1.%d.0", constants.MinimumControlPlaneVersion.Minor()+1),
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/config.go]: should be the same as:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants.go#L387

File: cmd/kubeadm/app/cmd/token.go [hunk 2]
```diff
@@ -208,14 +208,18 @@ func NewCmdTokenGenerate(out io.Writer) *cobra.Command {
...
-func RunCreateToken(out io.Writer, client clientset.Interface, cfgPath string, cfg *kubeadmapiv1beta2.InitConfiguration, printJoinCommand bool, kubeConfigFile string) error {
-	// KubernetesVersion is not used, but we set it explicitly to avoid the lookup
-	// of the version from the internet when executing LoadOrDefaultInitConfiguration
-	phaseutil.SetKubernetesVersion(&cfg.ClusterConfiguration)
+func RunCreateToken(out io.Writer, client clientset.Interface, cfgPath string, initCfg *kubeadmapiv1beta2.InitConfiguration, printJoinCommand bool, kubeConfigFile string) error {
+	// ClusterConfiguration is needed just for the call to LoadOrDefaultInitConfiguration
+	clusterCfg := &kubeadmapiv1beta2.ClusterConfiguration{
+		// KubernetesVersion is not used, but we set this explicitly to avoid
+		// the lookup of the version from the internet when executing LoadOrDefaultInitConfiguration
+		KubernetesVersion: constants.MinimumControlPlaneVersion.String(),
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/token.go]: possibly use this again here?
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants.go#L387

**Review by neolit123 — COMMENTED** [review]
thanks @rosti 
should we also clean this as part of this PR:
https://github.com/kubernetes/kubeadm/issues/1431
(possibly delete these tests?)

^ cc @fabriziopandini 
xref: https://github.com/kubernetes/kubernetes/pull/74797/files

**Comment by rosti** [comment]
> should we also clean this as part of this PR:
> [kubernetes/kubeadm#1431](https://github.com/kubernetes/kubeadm/issues/1431)
> (possibly delete these tests?)

This is outside of the scope of this PR. But I do envision a test grooming PR for the config tests after all the v1beta2 changes are merged. That can be part of it.

**Review by fabriziopandini — COMMENTED** [review]
@rosti thanks for this PR!
Just for confirmation, are you planning to change the internal API as well in as second stage?
I ask this because the disalignement introduced by this PR leads to some weird semantics that could create confusion in the long term e.g.
```
 internalcfg, err := configutil.DefaultedInitConfiguration(initCfg, clusterCfg)
```

File: cmd/kubeadm/app/cmd/alpha/kubeconfig.go [hunk 3]
```diff
@@ -96,9 +98,9 @@ func newCmdUserKubeConfig(out io.Writer) *cobra.Command {
 	}
 
 	// Add flags to the command
-	cmd.Flags().StringVar(&cfg.CertificatesDir, options.CertificatesDir, cfg.CertificatesDir, "The path where certificates are stored")
-	cmd.Flags().StringVar(&cfg.LocalAPIEndpoint.AdvertiseAddress, options.APIServerAdvertiseAddress, cfg.LocalAPIEndpoint.AdvertiseAddress, "The IP address the API server is accessible on")
-	cmd.Flags().Int32Var(&cfg.LocalAPIEndpoint.BindPort, options.APIServerBindPort, cfg.LocalAPIEndpoint.BindPort, "The port the API server is accessible on")
+	cmd.Flags().StringVar(&clusterCfg.CertificatesDir, options.CertificatesDir, clusterCfg.CertificatesDir, "The path where certificates are stored")
```
[the remark this answers is not in the document]
Comment (timothysc) [reply cmd/kubeadm/app/cmd/alpha/kubeconfig.go]: I'd hope we get that for free in the ComponentConfig Machinery in the long haul. 
/cc @luxas

**Review by timothysc — COMMENTED** [review]
Nothing stands out on initial inspection, are there any cases we can think about where we would want initconfig to force a change in clusterconfig?  I'd need to think about it.

File: cmd/kubeadm/app/cmd/alpha/kubeconfig.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rosti) [reply cmd/kubeadm/app/cmd/alpha/kubeconfig.go]: @fabriziopandini fixed
@timothysc that's one of the main goals of the component standard work ATM. There is even a proof of concept design on how to do that in #73494

**Comment by rosti** [comment]
> Just for confirmation, are you planning to change the internal API as well in as second stage?

@fabriziopandini indeed that's the case. The purpose for this PR is to allow v1beta2 release with 1.15 and that to be free from technical debt as much as possible. This will allow us to tackle the internal config in the next release cycle(s) via a series of consecutive PRs. Changes to the API will be made as part of this followup work and it will look way better.

**Comment by rosti** [comment]
> Nothing stands out on initial inspection, are there any cases we can think about where we would want initconfig to force a change in clusterconfig? I'd need to think about it.

@timothysc There is only one thing, that I can think of. `ClusterConfiguration.ControlPlaneEndpoint` can depend on `InitConfiguration.LocalAPIEndpoint` ATM. See [here](https://github.com/kubernetes/kubernetes/blob/fb41b7a801538c6c8c9385dc35bf1eaae92f06ab/cmd/kubeadm/app/util/endpoint.go#L31).

**Comment by timothysc** [comment]
> @timothysc There is only one thing, that I can think of. ClusterConfiguration.ControlPlaneEndpoint can depend on InitConfiguration.LocalAPIEndpoint ATM. See here. 

Blarg, can we truly separate then?

**Comment by fabriziopandini** [comment]
@rosti 
> ClusterConfiguration.ControlPlaneEndpoint can depend on InitConfiguration.LocalAPIEndpoint ATM. See here.

We can re-think this as a simplification of the single control-plane plane scenario: controlPlaneEndpoint, if empty, get initialised with local API endpoint.

Shifting from the current behaviour to a simple "defaulting" rule will remove the dependency and probably also cleanup some code
wdyt?

More generally speaking, controlPlaneEndpoint (always one) and local API endpoints (potentially many) were splitted intentionally

**Comment by rosti** [comment]
@timothysc @fabriziopandini yes, we can simplify things there. We need to dynamically default `ControlPlaneEndpoint` to the `LocalAPIEndpoint` even for the single control plane case.
We also have to ensure that we do the same on upgrade too.

However, this is outside of the scope of this PR. I can do a followup today or tomorrow.

**Comment by rosti** [comment]
@timothysc @fabriziopandini after looking at this I reached the following conclusions:

1. We should consider only `GetControlPlaneEndpoint` as it's the only place where `ControlPlaneEndpoint` and `LocalAPIEndpoint` depend on each other.
2. `GetControlPlaneEndpoint` is (almost) always used in a context where we are joining a control plane or have a control plane. This implies, that `LocalAPIEndpoint` is always present, either from Init or Join configs or reconstructed from the config map.
3. Only `kubeadm alpha kubeconfig user` uses `GetControlPlaneEndpoint` when not having a control plane, but there the usage is sort of a hack to get a control plane URL from the `LocalAPIEndpoint` (supplied at the command line).

At this point I think, that it's safe to leave things as they are. If we need we can address this later on.

Either way, the steps to get always something inside `ControlPlaneEndpoint` are the following:
- Make sure that this field is always filled in during kubeadm init - even for the single control plane case. The existing logic in `GetControlPlaneEndpoint` can be used for that. This needs to be done in the early stages of init, before any phases are executed and, preferably, just after the config is loaded.
- Make sure that the init upload-config phase does the same thing (in case it's called standalone).
- Make sure that upon upgrade this field is filled in. If it's empty, then the usual case is a single control plane cluster so filling it in won't be a big deal.
- Make sure that `kubeadm config upload` functionality will require complicating a few things to ensure a proper `ControlPlaneEndpoint` is uploaded.
- Replace `GetControlPlaneEndpoint` with `GetControlPlaneURL` that uses `ClusterConfiguration.ControlPlaneEndpoint`, performs a bunch of checks on it and returns an HTTPS URL.
- Add a check to config upload APIs.
- Some tests and lots of testing.

One strange thing, that I noticed in `GetControlPlaneEndpoint`, is that if `ControlPlaneEndpoint` has only a host name, but not a port in it, then the port is taken from the `LocalAPIEndpoint`. This means, that one can override the port number of the API server gateway on a per-control-plane basis. This case is not covered by the above steps as they require fixed port number in `GetControlPlaneEndpoint` or for the port number to be defaulted. I cannot think of a valid case where such port redirection would be useful for someone.

Anyway, it's late in the cycle, so I think, that we are safe with doing nothing. I have some partial code implementing the above schema, but let's keep it shelved for now and only proceed with this if there is no other option.

**Comment by fabriziopandini** [comment]
@rosti thanks for the detailed info
+1 for me on going on without changes and logging an issue for executing on "getting always something in `controlPlaneAddres`" next cycle

> I noticed in GetControlPlaneEndpoint, is that if ControlPlaneEndpoint has only a hostname, but not a port in it

this is something that probably should be fixed as well

**Review by neolit123 — COMMENTED** [review]
/lgtm
/hold

**Review by fabriziopandini — APPROVED** [review]
@rosti, Great job!
/approve 
/lgtm

**Comment by neolit123** [comment]
@rosti the verify failures seem adequate:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/77739/pull-kubernetes-verify/1133439642960400385

**Review by fabriziopandini — APPROVED** [review]
Thanks @rosti 
/lgtm

**Pull request merged**